Feature/container

mmiesch opened the pull request:
This PR adapts the jedi-stack build system for use in building the containers.  For now, the modules capability of jedi-stack is not used for the containers: everything is installed in /usr/local.  This is because our working model is to have one compiler/mpi combination for each container, so a single software stack.  If our use cases change in the future we can easily change this to install modules in the containers.

aerorahul commented on modulefiles/compiler/compilerName/compilerVersion/jedi-openmpi/jedi-openmpi.lua:
```diff
@@ -26,6 +26,9 @@ prepend_path("MODULEPATH", mpath)
 setenv("FC",  "mpifort")
```
`FC`, `CC`, `CXX`, should not be MPI versions in the modulefile.

mmiesch replied:
Thanks @aerorahul for your comments.

Regarding question 1, there are several reasons to stick to one container per compiler/mpi combination.  First, we do not have a license to distribute intel compilers or impi in a container.  In the future we will build application containers that have intel-build binaries and runtime libraries but not the compilers.   Second, the gnu/openmpi stack should be sufficient for the vast majority of JEDI users (over 90%?).  Including a gnu/mpich tree would roughly double the size of the container, making it less convenient for that majority.   More serious JEDI users who want to test multiple configurations may use other platforms such as AWS or HPC environment modules.  As I mentioned in the PR, if there is a need for containers with multiple modules, we can add that capability in the future.  But, my impression is that that is not a common approach in the HPC container community.

Regarding questions 2-3, my approach was to define MPI_ and SERIAL_ versions of the compilers in the modules and let the build scripts define FC, CC, and CX before calling cmake or configure, choosing either the SERIAL or the MPI versions as appropriate.  In my latest experience with eckit, I find that it does not compile correctly unless I set FC, CC, and CXX to be mpifort, mpicc, and mpicxx.  And I don't see why one should not define the MPI_FC, MPI_CC, and MPI_CXX in the modulefiles.  The modules define the executables, e.g. mpifort, so why not also define the environment variables that point to them?

aerorahul commented on modulefiles/compiler/compilerName/compilerVersion/jedi-mpich/jedi-mpich.lua:
```diff
@@ -27,6 +27,10 @@ setenv("FC",  "mpifort")
 setenv("CC",  "mpicc")
 setenv("CXX", "mpicxx")
 
+setenv("MPI_FC",  "mpifort")
```
`mpifort` is not a valid compiler in the Intel family.
This is mixing compiler families.

mmiesch replied:
I'm not sure what you mean here - do you mean that MPI_FC is a variable used by intel?  The way I am using it has no relation to intel.  I'm just defining it here and then setting, e.g. FC=$MPI_FC in scripts that need the wrappers.  This should work for any compiler/mpi combination, provided that you only have one combination loaded.

aerorahul replied:
the templated module file is under `compilerName`, which if it is intel, `mpifort` is incorrect.
You may have to create a different tree `modulefiles/compiler/intel/...` for `intel`.

mmiesch replied:
Thanks @aerorahul - I'm working on your comments now - quick question - I thought mpifort was the wrapper defined by mpich, not by the compiler.  So, if you were to build mpich with intel, wouldn't the wrapper still be called mpifort (in this case wrapped around ifort)?   If you were to use intel mpi it would be different, mpiifort, but this is taken into account in the modulefiles/compiler.../jedi-impi.lua file.

aerorahul commented on modulefiles/compiler/compilerName/compilerVersion/jedi-mpich/jedi-mpich.lua:
```diff
@@ -27,6 +27,10 @@ setenv("FC",  "mpifort")
 setenv("CC",  "mpicc")
```
`FC`, `CC`, `CXX`, should not be MPI versions in the modulefile.
It is the job of the build script or `cmake` to find the appropriate compiler wrappers for MPI.

mmiesch replied:
But then wouldn't each build script have to include a case statement that defines the various options, e.g. FC = mpifort, mpif90, or mpiifort, depending on your platform?  Why not just define MPI_FC once and write the build scripts to use that?

aerorahul replied:
Ok. I am convinced on this.
In the meta-module for jedi, if you could retain `FC`, `CC`, `CXX` and add `SERIAL_` and `MPI_`, that is fine.

aerorahul commented on modulefiles/core/jedi-intel/jedi-intel.lua:
```diff
@@ -18,9 +18,9 @@ local opt = os.getenv("OPT") or "/opt/modules"
-setenv("FC",  "ifort")
-setenv("CC",  "icc")
-setenv("CXX", "icpc")
+setenv("SERIAL_FC",  "ifort")
```
Retain `FC`, `CC`, `CXX`.
`cmake` relies on these to determine the compiler.
`SERIAL_CC`, and `MPI_CC`, are not standard way's of defining a compiler environment variable.
They can be used in a script, but do not belong in modulefiles.

mmiesch replied:
Yes, I understand that these variables are not recognized by CMake but they are understood by the build scripts.  For example, in the lapack build script I set FC=$SERIAL_FC before calling cmake while in the odb build script I set FC=$MPI_FC before calling ecbuild.

aerorahul commented on modulefiles/mpi/compilerName/compilerVersion/mpiName/mpiVersion/netcdf/netcdf.lua:
```diff
@@ -26,6 +26,7 @@ prepend_path("DYLD_LIBRARY_PATH", pathJoin(base,"lib"))
 prepend_path("CPATH", pathJoin(base,"include"))
 prepend_path("MANPATH", pathJoin(base,"share","man"))
 
+setenv("NETCDF", base)
```
Is there a reason `NETCDF_ROOT` is not sufficient?
`cmake` looks for `NETCDF_ROOT` as is standard.
If an application is looking for `NETCDF`, that should be obtained from `NETCDF_ROOT` in the build script of that application.

mmiesch replied:
I get the impression that setting the NETCDF variable is a standard practice when installing netcdf.  For example, when you load a netcdf module on Cheyenne, the NETCDF variable is set.  Why not set both and avoid the redundancy of setting it in the buildscripts (similarly for PNETCDF)?

aerorahul replied:
`Cheyenne` does it for convenience. Not true on `Theia`, `Discover`, `S4`, `Gaea`
But fine.

aerorahul commented on modulefiles/mpi/compilerName/compilerVersion/mpiName/mpiVersion/pnetcdf/pnetcdf.lua:
```diff
@@ -24,6 +24,7 @@ prepend_path("LIBRARY_PATH", pathJoin(base,"lib"))
 prepend_path("CPATH", pathJoin(base,"include"))
 prepend_path("MANPATH", pathJoin(base,"share","man"))
 
+setenv("PNETCDF", base)
```
Same comment as `NETCDF` for `NETCDF_ROOT`
The user environment is being inundated with too many variables that do the same thing.

mmiesch replied:
If multiple packages use $PNETCDF, I don't see the harm.

aerorahul commented on buildscripts/build_stack.sh:
```diff
@@ -35,24 +41,38 @@ else
+
 fi    
 
+export CC=$MPI_CC
```
This is not true when building serial applications; e.g. `szip`, `udunits`, `zlib` etc.

aerorahul commented on buildscripts/libs/build_eccodes.sh:
```diff
@@ -8,14 +8,30 @@ version=$1
+    fi
+
+else
+    prefix="/usr/local"
```
Is this really needed? This defeats the point of providing `PREFIX` from above.

mmiesch replied:
You're right - it's not really needed if /usr/local is set in the config file.  But, hardwiring it in like this provides a failsafe for improperly setting the config file.  If this is not here, and if a developer were to change the PREFIX in the config file, then the container would not build properly.  This is just another way to make the container building virtually automatic.

aerorahul commented on buildscripts/libs/build_eccodes.sh:
```diff
@@ -8,14 +8,30 @@ version=$1
+    prefix="/usr/local"
+fi
+
+export FC=$MPI_FC
```
ecCodes relies on `ecbuild`.
ecbuild finds MPI wrappers. `FC, CC, CXX` Besides, mpi is not being loaded, so `MPI_FC` is coming from?

aerorahul commented on buildscripts/libs/build_eckit.sh:
```diff
@@ -9,31 +9,39 @@ version=$1
+    fi
+
+else
+    prefix="/usr/local"
```
Same comment as in ecCodes.

aerorahul commented on buildscripts/libs/build_eckit.sh:
```diff
@@ -9,31 +9,39 @@ version=$1
+    prefix="/usr/local"
+fi
+
+export FC=$MPI_FC
```
Same comment as in `ecCodes`

aerorahul commented on buildscripts/libs/build_eckit.sh:
```diff
@@ -9,31 +9,39 @@ version=$1
+export CXX=$MPI_CXX
 
 export F9X=$FC
 export FFLAGS="-fPIC"
```
`ecbuild` finds the appropriate `FFLAGS`, `CFLAGS` etc. They don't need to be specified.
They are required/necessary for builds using `configure`.

mmiesch replied:
Thanks - good to know - I'll take those out

aerorahul commented on buildscripts/libs/build_eigen.sh:
```diff
@@ -8,12 +8,23 @@ version=$1
+    set +x
+    source $MODULESHOME/init/bash
+    module load jedi-$COMPILER
+    module load boost-headers
```
`Boost` is not required for Eigen.
If it is, then the modulefile for `Eigen` should be updated to load it as a pre-requisite.

mmiesch replied:
I think I did this for a reason - it's not required but I seem to recall it was going to use them if they were there.  AWS is unavailable right now but I'll double-check this when it comes back up.  Either way, I'll modify the modulefile - thanks!

mmiesch replied:
Hi @aerorahul - so I checked on this eigen/boost dependency issue.  If I compile without loading the boost-headers module, the build log does still find and (apparently) use boost (maybe it comes with its own version of some of the header files??):
~~~~~
> grep -i boost eigen3_noboost.log 
-- Boost version: 1.54.0
-- Enabled backends:  Qt4 support,  Boost.Multiprecision, 
~~~~~
If I compile with the boost-header module loaded, I get this:
~~~~~
> grep -i boost eigen3.log 
  1) git/2.10.2   2) gnu/7.3.0   3) jedi-gnu/7.3.0   4) openmpi/3.1.3   5) jedi-openmpi/3.1.3   6) boost-headers/1.68.0
-- Boost version: 1.68.0
-- Enabled backends:  Qt4 support,  Boost.Multiprecision, 
~~~~~
So it is finding and apparently using the more up-to-date headers in the boost module.  I'm not sure what it is used for - maybe just multiprecision, maybe more.  I will put it into the eigen modulefile as an optional package - load it if it is there.

mmiesch replied:
Actually - I don't see an obvious way in the lmod documentation to do an optional load - @aerorahul - do you know how to tell the eigen3 modulefile to load boost-headers if it is there but if it is not, then proceed anyway?

aerorahul commented on buildscripts/libs/build_eigen.sh:
```diff
@@ -8,12 +8,23 @@ version=$1
+if $MODULES; then
+    set +x
+    source $MODULESHOME/init/bash
+    module load jedi-$COMPILER
```
`Eigen` is all headers. No compiler is required here.

mmiesch replied:
I believe it runs unit tests as part of the normal make/make install - from the log file:
~~~~~
-- Check for working CXX compiler: /glade/u/apps/ch/opt/gnu/7.3.0/bin/g++
-- Check for working CXX compiler: /glade/u/apps/ch/opt/gnu/7.3.0/bin/g++ -- works
[...]
-- ************************************************************
-- ***    Eigen's unit tests configuration summary          ***
-- ************************************************************
[...]
 CXX_VERSION:       g++ (GCC) 7.3.0
 CXX_FLAGS:          -pedantic -Wall -Wextra -Wundef -Wcast-align -Wchar-subscripts -Wnon-virtual-dtor -Wunused-local-typedefs -Wpointer-arith -Wwrite-strings -Wformat-security -Wlogical-op -Wdouble-promotion -Wno-psabi -Wno-variadic-macros -Wno-long-long -fno-check-new -fno-common -fstrict-aliasing -ansi
~~~~~

aerorahul commented on buildscripts/libs/build_lapack.sh:
```diff
@@ -30,22 +43,15 @@ url="http://www.netlib.org/lapack/$software.tgz"
 # Add CMAKE_INSTALL_LIBDIR to make sure it will be installed under lib not lib64
 cmake -DCMAKE_INSTALL_PREFIX=$prefix -DCMAKE_BUILD_TYPE=Release -DCMAKE_INSTALL_LIBDIR:PATH=$prefix/lib \
-      -DCMAKE_Fortran_COMPILER=$FC -DCMAKE_Fortran_FLAGS=$FCFLAGS ..
+      -DCMAKE_Fortran_COMPILER=$SERIAL_FC -DCMAKE_Fortran_FLAGS=$FCFLAGS ..
```
`cmake` finds the appropriate fortran compiler based on `$FC` in the environment variable. This is not necessary.

aerorahul commented on modulefiles/core/jedi-gnu/jedi-gnu.lua:
```diff
@@ -18,9 +18,9 @@ local opt = os.getenv("OPT") or "/opt/modules"
-setenv("FC",  "gfortran")
-setenv("CC",  "gcc")
-setenv("CXX", "g++")
+setenv("SERIAL_FC",  "gfortran")
```
please retain `FC`, `CC`, `CXX`.
These are standard environment variables associated with a compiler.
`cmake` relies on these to determine the compiler and its family.
`SERIAL_FC` is not a standard, and a convenience variable.  It should be inferred from `$FC` in the environment when required.

mmiesch replied:
If SERIAL_FC is used in a consistent way by the build scripts I don't see the harm.  I find that eckit and ufo-bundle do not compile for me unless I set FC=mpifort.  But this would be inappropriate for lapack, which just uses the serial compiler.  It may be that CMake "should" figure all this out but in my experience it does not always succeed in doing so - why not help it out by telling it explicitly what compiler you want it to use?

aerorahul replied:
I see. I never set `FC=mpif90`, etc and eckit, ufo-bundle etc work.
In their cmakelists.txt, they should have ecbuild_find_mpi and then LINK_LIBRARIES(MPI_C_...)
May be it is an issue there.

mmiesch commented:
I'm going to close this PR for now to address Rahul's comments - I'll re-open it after a few commits

mmiesch commented:
Ok - I'm re-opening this PR.  I think I addressed all of @aerorahul's comments with code changes and/or responses.  Please let me know if I missed anything.  In the process, I also fixed a few bugs and made some minor changes to get this working on Cheyenne.

I chose to keep the SERIAL_FC/MPI_FC approach (also CC, CXX) because it works for me on multiple platforms and because, in my opinion, it is nicely transparent - the developer can see which compiler (serial or parallel) is being used in each of the build scripts.  However, I did add the FC, CC, and CXX flags back into the compiler modulefiles for use with CMake as Rahul requested.  I also chose to keep the NETCDF and PNETCDF variables in the modulefiles - it seems like multiple packages look for these variables so I believe it is beneficial to define them, though I do appreciate the goal of not defining too many gratuitous environment variables unless they are actually used (and I believe these are).  As Rahul suggested, I omitted the CFLAGS, FCFLAGS, CXXFLAGS in repos that are built with ecbuild.  Everything works on Cheyenne.

aerorahul commented:
Thanks @mmiesch.
All changes look good.
